### ISSUE TITLE
Unexpected query planner behavior with fragments within an union

### ISSUE THREAD

**Issue by utay** [apollographql/router#3172]
I'm using Apollo router 1.19.1 and would like to report 2 issues that are probably related.

#### 1. The query planner fails unexpectedly with the following error:
{
      "message": "value retrieval failed: router bridge error: the deno runtime raised an error: `request: couldn't receive response couldn't deserialize payload `12356547438014710016`: `deno: couldn't deserialize response : `Error(\"invalid neither null nor empty object: found Object {\\\"code\\\": String(\\\"QUERY_PLANNING_FAILED\\\"), \\\"exception\\\": Object {\\\"stacktrace\\\": Array [String(\\\"Cannot add selection of field \\\\\\\"OfferNotification.user\\\\\\\" to selection set of parent type \\\\\\\"Notification\\\\\\\"\\\")]}}\", line: 0, column: 0)``.`",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }

<details>
<summary>Minimal supergraph to reproduce</summary>

```
schema
  @link(url: "https://specs.apollo.dev/link/v1.0")
  @link(url: "https://specs.apollo.dev/join/v0.3", for: EXECUTION)
{
  query: Query
}

directive @join__enumValue(graph: join__Graph!) repeatable on ENUM_VALUE

directive @join__field(graph: join__Graph, requires: join__FieldSet, provides: join__FieldSet, type: String, external: Boolean, override: String, usedOverridden: Boolean) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION

directive @join__graph(name: String!, url: String!) on ENUM_VALUE

directive @join__implements(graph: join__Graph!, interface: String!) repeatable on OBJECT | INTERFACE

directive @join__type(graph: join__Graph!, key: join__FieldSet, extension: Boolean! = false, resolvable: Boolean! = true, isInterfaceObject: Boolean! = false) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR

directive @join__unionMember(graph: join__Graph!, member: String!) repeatable on UNION

directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA

type CardNotification implements NotificationInterface
  @join__implements(graph: BACKEND, interface: "NotificationInterface")
  @join__type(graph: BACKEND)
{
  id: Int!
}

type CurrentUser implements PublicUserInfoInterface
  @join__implements(graph: BACKEND, interface: "PublicUserInfoInterface")
  @join__type(graph: BACKEND)
{
  id: Int!
  slug: String!
  notifications: NotificationConnection!
}

scalar join__FieldSet

enum join__Graph {
  BACKEND @join__graph(name: "backend", url: "https://api.sorare.dev/graphql")
}

scalar link__Import

enum link__Purpose {
  """
  `SECURITY` features provide metadata necessary to securely resolve fields.
  """
  SECURITY

  """
  `EXECUTION` features provide metadata necessary for operation execution.
  """
  EXECUTION
}

union Notification
  @join__type(graph: BACKEND)
  @join__unionMember(graph: BACKEND, member: "OfferNotification")
  @join__unionMember(graph: BACKEND, member: "CardNotification")
 = OfferNotification | CardNotification

type NotificationConnection
  @join__type(graph: BACKEND)
{
  nodes: [Notification!]!
}

interface NotificationInterface
  @join__type(graph: BACKEND)
{
  id: Int!
}

type OfferNotification implements NotificationInterface
  @join__implements(graph: BACKEND, interface: "NotificationInterface")
  @join__type(graph: BACKEND)
{
  id: Int!
  user: CurrentUser!
}

interface PublicUserInfoInterface
  @join__type(graph: BACKEND)
{
  id: Int!
  slug: String!
}

type Query
  @join__type(graph: BACKEND)
{
  currentUser: CurrentUser!
}

```
</details>

<details>
<summary>Query to reproduce</summary>

```
{
  currentUser {
    notifications {
      nodes {
        ... on NotificationInterface {
          id
        }
        ...Fragment1
        ...Fragment2
      }
    }
  }
}

fragment Fragment1 on OfferNotification {
  user {
    ... on CurrentUser {
      ...Fragment3
    }
  }
}

fragment Fragment2 on CardNotification {
  ...Fragment4
  id
}

fragment Fragment3 on PublicUserInfoInterface {
  slug
}

fragment Fragment4 on NotificationInterface {
  id
  __typename
}
```
</details>

#### 2. When I remove `__typename`  from `Fragment4`, the router sends this query to its subgraph:

<details>
<summary>Subgraph query</summary>

```
{
  currentUser {
    notifications {
      nodes {
        __typename
        ... on CardNotification {
          ...Fragment4
        }
        ... on NotificationInterface {
          __typename
        }
        ... on OfferNotification {
          user {
            slug
          }
        }
      }
    }
  }
}

fragment Fragment4 on NotificationInterface {
  id
}
```
</details>

which I think is wrong too because we loose `currentUser.notifications.nodes.id` if it's not a `CardNotification`

**Comment by utay** [comment]
FYI @abernix @pcmanus, it looks like a bug similar to https://github.com/apollographql/federation/issues/2592. Please feel free to move to the federation repo and let me know if you need more information!

**Comment by utay** [comment]
@pcmanus That's great to hear, thank you!!

**Comment by abernix** [comment]
@utay Thanks for reporting this.  Can you let us know if this is **not** fixed in https://github.com/apollographql/router/releases/tag/v1.20.0 which includes the PR referenced above (released in Federation 2.4.7)?

**Comment by utay** [comment]
Just wanted to confirm that router 1.20.0 / federation 2.4.7 fixed this issue, thank you!
